src/stats/mod.rs

```diff
@@ -0,0 +1,400 @@
+                crate = metriken
+            )]
+            pub static [<$ident _EX>]: Lazy<metriken::Counter> = metriken::Lazy::new(|| {
+                metriken::Counter::new()
+            });
+        }
+
+        paste! {
+            #[metriken::metric(
+                name = concat!($name, "/ok"),
+                crate = metriken
+            )]
+            pub static [<$ident _OK>]: Lazy<metriken::Counter> = metriken::Lazy::new(|| {
+                metriken::Counter::new()
+            });
+        }
+
+        paste! {
+            #[metriken::metric(
+                name = concat!($name, "/timeout"),
+                crate = metriken
+            )]
+            pub static [<$ident _TIMEOUT>]: Lazy<metriken::Counter> = metriken::Lazy::new(|| {
+                metriken::Counter::new()
+            });
+        }
+    }
+}
+
+heatmap!(
+    REQUEST_LATENCY,
+    "request_latency",
+    "distribution of request latencies in nanoseconds. incremented at time of requests disbatch."
+);
+
+heatmap!(
+    RESPONSE_LATENCY,
+    "response_latency",
+    "distribution of response latencies in nanoseconds. incremented at time of response receipt."
+);
+
+heatmap!(
+    SESSION_LIFECYCLE_REQUESTS,
+    "session_lifecycle_requests",
+    "distribution of requests per session lifecycle. incremented at time of session close."
+);
+
+heatmap!(PUBSUB_LATENCY, "pubsub_latency");
+
+heatmap!(PUBSUB_PUBLISH_LATENCY, "pubsub_publish_latency");
+
+gauge!(CONNECT_CURR, "client/connections/current");
+counter!(CONNECT_OK, "client/connect/ok");
+counter!(CONNECT_TIMEOUT, "client/connect/timeout");
+
+counter!(REQUEST, "client/request/total", "total requests dequeued");
+counter!(
+    REQUEST_OK,
+    "client/request/ok",
+    "requests that were successfully generated and sent"
+);
+counter!(
+    REQUEST_RECONNECT,
+    "client/connect/reconnect",
+    "requests to reconnect"
+);
+counter!(
+    REQUEST_UNSUPPORTED,
+    "client/request/unsupported",
+    "skipped requests due to protocol incompatibility"
+);
+
+counter!(
+    RESPONSE_EX,
+    "client/response/exception",
+    "responses which encountered some exception while processing"
+);
+counter!(
+    RESPONSE_RATELIMITED,
+    "client/response/ratelimited",
+    "backend indicated that we were ratelimited"
+);
+counter!(
+    RESPONSE_BACKEND_TIMEOUT,
+    "client/response/backend_timeout",
+    "responses indicating the backend timedout"
+);
+counter!(
+    RESPONSE_OK,
+    "client/response/ok",
+    "responses which were successful"
+);
+counter!(RESPONSE_TIMEOUT, "responses not received due to timeout");
+counter!(
+    RESPONSE_INVALID,
+    "responses that were invalid for the protocol"
+);
+
+counter!(RESPONSE_HIT, "client/response/hit");
+counter!(RESPONSE_MISS, "client/response/miss");
+
+// augment the get stats
+counter!(
+    GET_OK,
+    "client/request/get/ok",
+    "get requests that were successful"
+);
+counter!(
+    GET_TIMEOUT,
```

Comment:
Why do `get` and `set` get `*_TIMEOUT` metrics but not the other data structure related requests?

Reply:
The `request!()` macro is defining the base stat (eg: `HASH_GET`) plus the `_EX`, `_OK` and `_TIMEOUT` stats. `GET` and `SET` are handled differently because the protocol crate defines those metrics. A future change to the protocol crates in Pelikan so that we can opt-out of stats and define our own should allow us to make this more consistent.

Macro is defined here: https://github.com/iopsystems/rpc-perf/pull/9/files#diff-1317be48f7b8da36d198f84a4ac1ee6cbcede32d9604aef38b8212dec1d4c921R158